arroyo-controller/src/lib.rs

```diff
@@ -447,35 +447,69 @@ impl ControllerGrpc for ControllerServer {
             Status::unavailable(format!("Failed to connect to compiler service: {}", e))
         })?;
 
-        let udfs_rs = request.into_inner().udfs_rs.clone();
+        let definition = request.into_inner().definition.clone();
 
+        let mut function = None;
         {
-            let result = match parse_file(udfs_rs.as_str()) {
+            let result = match parse_file(definition.as_str()) {
                 Ok(result) => result,
                 Err(e) => {
                     return Ok(Response::new(CheckUdfsResp {
                         result: ValidationResult::Error as i32,
                         errors: vec![e.to_string()],
+                        udf_name: None,
                     }));
                 }
             };
 
             for item in result.items {
                 match item {
-                    Item::Fn(_) | Item::Use(_) => {}
+                    Item::Fn(f) => {
+                        if function.is_some() {
+                            return Ok(Response::new(CheckUdfsResp {
+                                result: ValidationResult::Error as i32,
+                                errors: vec!["Only one function is allowed in UDFs".to_string()],
+                                udf_name: None,
+                            }));
+                        }
+                        function = Some(f.sig.ident.to_string());
+                    }
+                    Item::Use(_) => {}
                     _ => {
                         return Ok(Response::new(CheckUdfsResp {
                             result: ValidationResult::Error as i32,
                             errors: vec![
                                 "Only functions and use statements are allowed in UDFs".to_string()
                             ],
+                            udf_name: None,
                         }))
                     }
                 }
             }
         }
 
-        client.check_udfs(CheckUdfsReq { udfs_rs }).await
+        // unwrap function or return error
+        let function = match function {
+            Some(f) => f,
+            None => {
+                return Ok(Response::new(CheckUdfsResp {
+                    result: ValidationResult::Error as i32,
+                    errors: vec!["No function found in UDF".to_string()],
+                    udf_name: None,
+                }))
+            }
+        };
```

Comment:
Let-else is really nice for this pattern:

```suggestion
        let Some(function) = function else {
              return Ok(Response::new(CheckUdfsResp {
                  result: ValidationResult::Error as i32,
                  errors: vec!["No function found in UDF".to_string()],
                  udf_name: None,
              }))
          }
        };
```

Reply:
Thanks, some day I'll get the hang of this.